cmd/stupid/main.go

```diff
@@ -20,60 +21,89 @@ func main() {
 			logger.SetLevel(lvl)
 		}
 	}
-	if len(os.Args) != 3 {
-		fmt.Printf("Usage: stupid config.yaml 500\n")
+	err := process(logger)
+	if err != nil {
+		fmt.Println(err)
 		os.Exit(1)
 	}
-	config := infra.LoadConfig(os.Args[1])
+	os.Exit(0)
+}
+
+func process(logger *log.Logger) error {
+	if len(os.Args) != 3 {
+		return errors.New(infra.UsageError)
+	}
 	N, err := strconv.Atoi(os.Args[2])
 	if err != nil {
-		panic(err)
+		return errors.New(infra.UsageError)
+	}
+	config, err := infra.LoadConfig(os.Args[1], logger)
+	if err != nil {
+		return err
+	}
+	crypto, err := config.LoadCrypto(logger)
+	if err != nil {
+		return err
 	}
-	crypto := config.LoadCrypto()
-
 	raw := make(chan *infra.Elements, 100)
 	signed := make([]chan *infra.Elements, len(config.Endorsers))
 	processed := make(chan *infra.Elements, 10)
 	envs := make(chan *infra.Elements, 10)
 	done := make(chan struct{})
-
+	ec := make(chan error)
```

Comment:
it seems to be a clever trick to reuse `errorCh` for errors and signals. However, there are several problems here:
- `errorCh` is now shared by several senders, and closed by *one* of them. This is a common panic caused by `send to closed channel`, and we are very likely to hit this with current implementation
- `errorCh` needs to be buffered so that no sender would be blocked
- it compromises readability to use single channel for distinct purposes, especially when there's no comment explaining this

a simple solution for the first problem is to *never* close channel - simple send a "special" error to signal the end, i.e.
```
if err == CompleteErr {
  // break to finish
} else if err != nil {
  // log error
}
```

**However, i would strongly recommend to use two separate channels for error and signal**, and you could easily drain with for-select loop

Reply:
Also, please use `errorC` instead of `ec` to be more explicit

Reply:
This is **NOT** suggested solution but just for the sake of argument, a simple solution for the first problem is to *never* close channel - simply send a "special" error to signal the end, i.e.
```
if err == CompleteErr {
  // break to finish
} else if err != nil {
  // log error
}
```

Reply:
Suppose `Observer.Start` returns a channel `finishCh`, and `errorCh` is used to process errors. You could use for-select to handle these two channels:
```
for {
  select {
  case err = <- errorCh:
    // process err, and exit(1) if necessary
  case <- finishCh:
    // equivalent to observer.Wait() has finished, break
  }
}
```
**or** start a separate goroutine to capture errors (and exit(1) if appropriate), so the `observer.Wait()` can remain mostly intact

Reply:
so if start a separate goroutine to capture errors and exit(1) ? when the separate goroutine exit with 1, the main goroutine and other goroutines will exit right?

Reply:
updated with for select

Reply:
> so if start a separate goroutine to capture errors and exit(1) ? when the separate goroutine exit with 1, the main goroutine and other goroutines will exit right?
correct (sorry for the delay)